File: 05-AI与机器学习/04-计算机视觉/Canny边缘检测/go/canny.go
// Canny 边缘检测的 Go 实现(与 python/canny.py 同题)。
//
// 文件划分:本文件只放流水线实现(高斯核 / Sobel / 量化 / NMS / 滞后阈值),
// 断言与入口在同包的 main.go 里 —— Go 单文件行数上限 300(OPTIMIZATION.md §1.1)。
//
// 运行:cd go && go run .
package main

import (
	"fmt"
	"math"
)

// sobelX / sobelY 是 OpenCV 文档给出的标准 3x3 Sobel 核。
var sobelX = [3][3]float64{{-1, 0, 1}, {-2, 0, 2}, {-1, 0, 1}}
var sobelY = [3][3]float64{{-1, -2, -1}, {0, 0, 0}, {1, 2, 1}}

// gaussian5x5Int 是 OpenCV 3.4 Canny 教程给出的整数核,除数为 159。
var gaussian5x5Int = [5][5]int{
	{2, 4, 5, 4, 2},
	{4, 9, 12, 9, 4},
	{5, 12, 15, 12, 5},
	{4, 9, 12, 9, 4},
	{2, 4, 5, 4, 2},
}

const gaussianDivisor = 159

// nmsOffsets 把量化后的 4 个梯度方向映射到该方向上要比较的两个邻居偏移。
// 索引 0/1/2/3 分别对应 0/45/90/135 度。
var nmsOffsets = [4][2][2]int{
	{{1, 0}, {-1, 0}},  // 0 度:水平梯度 -> 垂直边缘
	{{1, 1}, {-1, -1}}, // 45 度
	{{0, 1}, {0, -1}},  // 90 度:垂直梯度 -> 水平边缘
	{{-1, 1}, {1, -1}}, // 135 度
}

var total, passed int
var fails []string

func check(name string, cond bool, detail string) {
	total++
	if cond {
		passed++
		fmt.Printf("  [PASS] %s  %s\n", name, detail)
		return
	}
	fails = append(fails, name)
	fmt.Printf("  [FAIL] %s  %s\n", name, detail)
}

func clampInt(v, lo, hi int) int {
	if v < lo {
		return lo
	}
	if v > hi {
		return hi
	}
	return v
}

// correlate 通用奇数尺寸相关运算,clamp 复制边界(与 Python 版一致)。
func correlate(img [][]float64, kernel [][]float64) [][]float64 {
	h, w := len(img), len(img[0])
	k := len(kernel)
	half := k / 2
	out := make([][]float64, h)
	for y := 0; y < h; y++ {
		out[y] = make([]float64, w)
		for x := 0; x < w; x++ {
			acc := 0.0
			for ky := 0; ky < k; ky++ {
				yy := clampInt(y+ky-half, 0, h-1)
				for kx := 0; kx < k; kx++ {
					xx := clampInt(x+kx-half, 0, w-1)
					acc += kernel[ky][kx] * img[yy][xx]
				}
			}
			out[y][x] = acc
		}
	}
	return out
}

func gaussianKernel(size int, sigma float64) [][]float64 {
	half := size / 2
	flat := make([]float64, size)
	sum := 0.0
	for i := 0; i < size; i++ {
		d := float64(i - half)
		flat[i] = math.Exp(-(d * d) / (2 * sigma * sigma))
		sum += flat[i]
	}
	for i := range flat {
		flat[i] /= sum
	}
	k := make([][]float64, size)
	for i := 0; i < size; i++ {
		k[i] = make([]float64, size)
		for j := 0; j < size; j++ {
			k[i][j] = flat[i] * flat[j]
		}
	}
	return k
}

func toFloatGrid(m [3][3]float64) [][]float64 {
	out := make([][]float64, 3)
	for i := 0; i < 3; i++ {
		out[i] = []float64{m[i][0], m[i][1], m[i][2]}
	}
	return out
}

func blur(img [][]float64) [][]float64 { return correlate(img, gaussianKernel(5, 1.4)) }

func sobel(img [][]float64) ([][]float64, [][]float64) {
	return correlate(img, toFloatGrid(sobelX)), correlate(img, toFloatGrid(sobelY))
}

func magnitude(gx, gy [][]float64, l2 bool) [][]float64 {
	h, w := len(gx), len(gx[0])
	out := make([][]float64, h)
	for y := 0; y < h; y++ {
		out[y] = make([]float64, w)
		for x := 0; x < w; x++ {
			if l2 {
				out[y][x] = math.Hypot(gx[y][x], gy[y][x])
			} else {
				out[y][x] = math.Abs(gx[y][x]) + math.Abs(gy[y][x])
			}
		}
	}
	return out
}

// angle180 返回 [0,180) 的梯度方向(度)。
func angle180(gx, gy float64) float64 {
	a := math.Atan2(gy, gx) * 180.0 / math.Pi
	if a < 0 {
		a += 180.0
	}
	if a >= 180.0 {
		a -= 180.0
	}
	return a
}

// sector 把方向量化到 0/45/90/135,返回 0..3 的档位。
func sector(deg float64) int {
	switch {
	case deg < 22.5 || deg >= 157.5:
		return 0
	case deg < 67.5:
		return 1
	case deg < 112.5:
		return 2
	default:
		return 3
	}
}

// nonMax 沿梯度方向保留局部极大。epsRatio 是浮点残差护栏,0 表示关闭。
func nonMax(mag [][]float64, sec [][]int, epsRatio float64) [][]float64 {
	h, w := len(mag), len(mag[0])
	peak := 0.0
	for y := 0; y < h; y++ {
		for x := 0; x < w; x++ {
			if mag[y][x] > peak {
				peak = mag[y][x]
			}
		}
	}
	eps := peak * epsRatio
	out := make([][]float64, h)
	for y := 0; y < h; y++ {
		out[y] = make([]float64, w)
		for x := 0; x < w; x++ {
			m := mag[y][x]
			if m <= eps {
				continue
			}
			o := nmsOffsets[sec[y][x]]
			n1, n2 := 0.0, 0.0
			if xx, yy := x+o[0][0], y+o[0][1]; xx >= 0 && xx < w && yy >= 0 && yy < h {
				n1 = mag[yy][xx]
			}
			if xx, yy := x+o[1][0], y+o[1][1]; xx >= 0 && xx < w && yy >= 0 && yy < h {
				n2 = mag[yy][xx]
			}
			// 平局规则:先出现的邻居必须严格更小,后出现的允许相等。
			if m > n1 && m >= n2 {
				out[y][x] = m
			}
		}
	}
	return out
}

// hysteresis 双阈值 + 滞后,返回 0/1 边缘图。8 连通。
func hysteresis(mag [][]float64, low, high float64) [][]int {
	h, w := len(mag), len(mag[0])
	const weak, strong = 1, 2
	state := make([][]int, h)
	stack := make([][2]int, 0)
	for y := 0; y < h; y++ {
		state[y] = make([]int, w)
		for x := 0; x < w; x++ {
			if mag[y][x] >= high {
				state[y][x] = strong
				stack = append(stack, [2]int{x, y})
			} else if mag[y][x] >= low {
				state[y][x] = weak
			}
		}
	}
	for len(stack) > 0 {
		p := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		for dy := -1; dy <= 1; dy++ {
			for dx := -1; dx <= 1; dx++ {
				if dx == 0 && dy == 0 {
					continue
				}
				xx, yy := p[0]+dx, p[1]+dy
				if xx >= 0 && xx < w && yy >= 0 && yy < h && state[yy][xx] == weak {
					state[yy][xx] = strong
					stack = append(stack, [2]int{xx, yy})
				}
			}
		}
	}
	out := make([][]int, h)
	for y := 0; y < h; y++ {
		out[y] = make([]int, w)
		for x := 0; x < w; x++ {
			if state[y][x] == strong {
				out[y][x] = 1
			}
		}
	}
	return out
}

func canny(img [][]float64, low, high float64) [][]int {
	gx, gy := sobel(blur(img))
	mag := magnitude(gx, gy, false)
	h, w := len(gx), len(gx[0])
	sec := make([][]int, h)
	for y := 0; y < h; y++ {
		sec[y] = make([]int, w)
		for x := 0; x < w; x++ {
			sec[y][x] = sector(angle180(gx[y][x], gy[y][x]))
		}
	}
	return hysteresis(nonMax(mag, sec, 1e-9), low, high)
}

func stepImage(h, w, x0 int) [][]float64 {
	img := make([][]float64, h)
	for y := 0; y < h; y++ {
		img[y] = make([]float64, w)
		for x := 0; x < w; x++ {
			if x >= x0 {
				img[y][x] = 200
			}
		}
	}
	return img
}

func countEdges(e [][]int) int {
	n := 0
	for _, row := range e {
		for _, v := range row {
			n += v
		}
	}
	return n
}

File: 05-AI与机器学习/04-计算机视觉/Canny边缘检测/python/canny.py
# -*- coding: utf-8 -*-
"""Canny 边缘检测 — 五阶段流水线的最小实现(纯标准库)。

权威口径(docs.opencv.org 4.x `tutorial_py_canny`、OpenCV 3.4 `tutorial_canny_detector`、
Stanford EE368 讲义《Canny edge detector》,三者对同一算法的描述一致):

1. **去噪**:5x5 高斯滤波。3.4 教程给出除数 159 的整数核(本文件 GAUSSIAN_5X5_INT)。
2. **求梯度**:一对 3x3 Sobel 核;幅值 `G = sqrt(Gx^2+Gy^2)`(cv.Canny 的 `L2gradient=True`)
   或近似 `G = |Gx| + |Gy|`(默认 `L2gradient=False`)。
3. **非极大值抑制**:梯度方向量化到 0/45/90/135 四个角度,与该方向上两个邻居比较,
   只留局部极大 —— 结果是「细边」二值图。
4. **双阈值 + 滞后**:`> high` 必为边;`< low` 必非边;落在两者之间**只有与确定边连通**才保留。
5. **阈值比**:Canny 建议 high:low 在 **2:1 ~ 3:1** 之间(OpenCV 教程的 trackbar 例子用 ratio=3)。

工程取舍(与 OpenCV 的差异,已在 README「注意事项」中写明):
- 边界用 **clamp 复制**,OpenCV 默认是 `BORDER_DEFAULT`(reflect_101);差异只出现在图像
  最外 1~2 像素,不影响内部结构。
- 只实现 `apertureSize=3`(OpenCV 还支持 5/7,由 getDerivKernels 生成)。
- 滞后连通性用 **8 连通**(OpenCV 教程原文 "connected to sure-edge pixels" 未限定邻域)。
"""
import math

# OpenCV Canny 注释里给出的标准 3x3 Sobel 核(docs.opencv.org Sobel Derivatives 同款)
SOBEL_X = ((-1, 0, 1), (-2, 0, 2), (-1, 0, 1))
SOBEL_Y = ((-1, -2, -1), (0, 0, 0), (1, 2, 1))

# OpenCV 3.4 Canny 教程原文:K = 1/159 * [[2,4,5,4,2],[4,9,12,9,4],[5,12,15,12,5],...]
GAUSSIAN_5X5_INT = (
    (2, 4, 5, 4, 2),
    (4, 9, 12, 9, 4),
    (5, 12, 15, 12, 5),
    (4, 9, 12, 9, 4),
    (2, 4, 5, 4, 2),
)
GAUSSIAN_5X5_DIVISOR = 159

# 非极大值抑制:把梯度方向量化成 4 档后,沿该方向取的两个邻居偏移
# (图像坐标 y 向下;角度以 +x 为 0 逆时针到 y 轴为 90,即 atan2(gy, gx))
NMS_OFFSETS = {
    0: ((1, 0), (-1, 0)),      # 水平梯度 -> 垂直边缘
    45: ((1, 1), (-1, -1)),
    90: ((0, 1), (0, -1)),     # 垂直梯度 -> 水平边缘
    135: ((-1, 1), (1, -1)),
}


def gaussian_kernel_2d(size=5, sigma=1.4):
    """归一化的 2D 高斯核(均值 0,标准差 sigma)。size 必须为奇数。"""
    if size % 2 == 0:
        raise ValueError("kernel size must be odd")
    half = size // 2
    flat = [math.exp(-(i * i) / (2.0 * sigma * sigma)) for i in range(-half, half + 1)]
    total = sum(flat)
    flat = [v / total for v in flat]
    return tuple(tuple(fy * fx for fx in flat) for fy in flat)


def correlate(img, kernel):
    """通用奇数尺寸相关运算(clamp 边界),kernel 为等宽方形元组。"""
    h, w = len(img), len(img[0])
    k = len(kernel)
    half = k // 2
    out = [[0.0] * w for _ in range(h)]
    for y in range(h):
        oy = out[y]
        for x in range(w):
            acc = 0.0
            for ky in range(k):
                yy = y + ky - half
                yy = 0 if yy < 0 else (h - 1 if yy >= h else yy)
                krow = kernel[ky]
                irow = img[yy]
                for kx in range(k):
                    xx = x + kx - half
                    xx = 0 if xx < 0 else (w - 1 if xx >= w else xx)
                    acc += krow[kx] * irow[xx]
            oy[x] = acc
    return out


def blur(img, size=5, sigma=1.4):
    """5x5 高斯平滑(第 1 阶段:去噪)。"""
    return correlate(img, gaussian_kernel_2d(size, sigma))


def sobel_gradients(img, aperture=3):
    """第 2 阶段:Gx / Gy。本实现只支持 aperture=3(OpenCV 还支持 5/7)。"""
    if aperture != 3:
        raise ValueError("this demo implements apertureSize=3 only")
    return correlate(img, SOBEL_X), correlate(img, SOBEL_Y)


def gradient_magnitude(gx, gy, l2gradient=False):
    """L2gradient=True 走 sqrt(Gx^2+Gy^2);默认 False 走 |Gx|+|Gy|(OpenCV 教程)。"""
    h, w = len(gx), len(gx[0])
    if l2gradient:
        return [[math.hypot(gx[y][x], gy[y][x]) for x in range(w)] for y in range(h)]
    return [[abs(gx[y][x]) + abs(gy[y][x]) for x in range(w)] for y in range(h)]


def gradient_angle(gx, gy):
    """梯度方向,归一到 [0,180):方向与方向+180 是同一条直线,无需区分。"""
    a = math.degrees(math.atan2(gy, gx))
    if a < 0.0:
        a += 180.0
    if a >= 180.0:
        a -= 180.0
    return a


def direction_sector(angle_deg):
    """把 [0,180) 量化到 4 个方向之一:0 / 45 / 90 / 135(OpenCV 3.4 教程原文)。"""
    if angle_deg < 22.5 or angle_deg >= 157.5:
        return 0
    if angle_deg < 67.5:
        return 45
    if angle_deg < 112.5:
        return 90
    return 135


def non_max_suppression(mag, sectors, eps_ratio=1e-9):
    """第 3 阶段:沿梯度方向只保留局部极大。

    平局规则必须显式写死:理想阶跃边缘两侧的梯度幅值**在数学上严格相等**
    (见 README「注意事项」),用 `>=` 比较会留下两列、用 `>` 比较会一列不留。
    这里取「先出现的邻居(NMS_OFFSETS 的次序,即 +1 侧)必须严格更小、后出现的
    邻居允许相等」,保证任何情况下每行只留 1 列。

    `eps_ratio` 是**浮点残差护栏**:平坦区的 |Gx|、|Gy| 因累加顺序不同会留下
    1e-14 量级的残差(实测 2.84e-14),用 `m == 0.0` 判断"无梯度"会被它骗过,
    于是在纯平坦区留下伪边。这里按"地图峰值 × eps_ratio"设一个相对下限;
    真实管线里这一层由双阈值兜住,但 NMS 自身不该输出噪声。取 0 可关掉护栏。
    """
    h, w = len(mag), len(mag[0])
    peak = max(v for row in mag for v in row)
    eps = peak * eps_ratio
    out = [[0.0] * w for _ in range(h)]
    for y in range(h):
        for x in range(w):
            m = mag[y][x]
            if m <= eps:
                continue
            (d1x, d1y), (d2x, d2y) = NMS_OFFSETS[sectors[y][x]]
            n1 = n2 = 0.0
            if 0 <= x + d1x < w and 0 <= y + d1y < h:
                n1 = mag[y + d1y][x + d1x]
            if 0 <= x + d2x < w and 0 <= y + d2y < h:
                n2 = mag[y + d2y][x + d2x]
            if m > n1 and m >= n2:
                out[y][x] = m
    return out


def hysteresis(mag, low, high):
    """第 4 阶段:双阈值 + 滞后。返回 0/1 边缘图。

    OpenCV 教程的判据:> high 为「确定边」;< low 直接丢弃;两者之间的像素
    只有**与确定边连通**才算边,否则丢弃(这就是"滞后")。
    """
    if low > high:
        raise ValueError("low threshold must not exceed high threshold")
    h, w = len(mag), len(mag[0])
    WEAK, STRONG = 1, 2
    state = [[0] * w for _ in range(h)]
    stack = []
    for y in range(h):
        for x in range(w):
            v = mag[y][x]
            if v >= high:
                state[y][x] = STRONG
                stack.append((x, y))
            elif v >= low:
                state[y][x] = WEAK
    while stack:
        x, y = stack.pop()
        for dy in (-1, 0, 1):
            for dx in (-1, 0, 1):
                if dx == 0 and dy == 0:
                    continue
                xx, yy = x + dx, y + dy
                if 0 <= xx < w and 0 <= yy < h and state[yy][xx] == WEAK:
                    state[yy][xx] = STRONG
                    stack.append((xx, yy))
    return [[1 if v == STRONG else 0 for v in row] for row in state]


def canny(img, low, high, aperture=3, l2gradient=False, sigma=1.4):
    """完整流水线:平滑 -> 梯度 -> NMS -> 滞后阈值。返回 0/1 边缘图。"""
    src = blur(img, 5, sigma)
    gx, gy = sobel_gradients(src, aperture)
    h, w = len(gx), len(gx[0])
    mag = gradient_magnitude(gx, gy, l2gradient)
    sectors = [
        [direction_sector(gradient_angle(gx[y][x], gy[y][x])) for x in range(w)]
        for y in range(h)
    ]
    return hysteresis(non_max_suppression(mag, sectors), low, high)


def auto_thresholds(img, sigma=0.33):
    """中位数启发式阈值(工程惯例,非 Canny 原文/Canny 官方口径)。

    low = (1-sigma)*median, high = (1+sigma)*median —— 目的是省掉逐图调参;
    与 cv.Canny 一样,阈值作用于**梯度幅值**而非灰度,这里仅按灰度中位数标定尺度。
    """
    vals = sorted(v for row in img for v in row)
    n = len(vals)
    med = vals[n // 2] if n % 2 else 0.5 * (vals[n // 2 - 1] + vals[n // 2])
    return (1.0 - sigma) * med, (1.0 + sigma) * med
